lib/metadata/in_memory/backend.js

```diff
@@ -83,6 +83,27 @@ const metastore = {
                 if (err) {
                     return cb(err);
                 }
+                /*
+                    valid combinations of versioning options:
+                    - !versioning && !versionId: normal non-versioning put
+                    - versioning && !versionId: create a new version
+                    - versionId: update (PUT/DELETE) an existing version,
+                        and also update master version in case the put
+                        version is newer or same version than master.
+                        if versionId === '' update master version
+                */
+
+                if (params && params.versionId) {
+                    objVal.versionId = params.versionId; // eslint-disable-line
+                    const mst = metadata.keyMaps.get(bucketName).get(objName);
+                    if (mst && mst.versionId === params.versionId || !mst) {
+                        metadata.keyMaps.get(bucketName).set(objName, objVal);
+                    }
+                    // eslint-disable-next-line
+                    objName = formatVersionKey(objName, params.versionId);
+                    metadata.keyMaps.get(bucketName).set(objName, objVal);
+                    return cb(null, `{"versionId":"${objVal.versionId}"}`);
+                }
                 if (params && params.versioning) {
```

Comment:
Just checking, should `params.versionId === ''` be moved above this check?  if `params.versionId === ''` and `params.versioning`

Reply:
If versioning is true, then we expect either no version specified (undefined) or a valid versionId. The next one is for versioning suspended buckets I believe. It should be ok for that case.